File: exam_prep/IST603/assignment/part4/FindMaximum.java
package exam_prep.IST603.assignment.part4;

/**
 * Generic method findMaximum() with bounded type parameters.
 * Accepts an array of numbers (Integer, Double, etc.) and returns the maximum value.
 */
public class FindMaximum {

    /**
     * Bounded type parameter: T must extend Number and be Comparable with itself,
     * so only numeric types (Integer, Double, Long, etc.) are allowed.
     */
    public static <T extends Number & Comparable<T>> T findMaximum(T[] array) {
        if (array == null || array.length == 0) {
            throw new IllegalArgumentException("Array must not be null or empty");
        }
        T max = array[0];
        for (int i = 1; i < array.length; i++) {
            if (array[i].compareTo(max) > 0) {
                max = array[i];
            }
        }
        return max;
    }

    public static void main(String[] args) {
        Integer[] ints = { 3, 7, 2, 9, 1 };
        System.out.println("Max Integer: " + findMaximum(ints));  // 9

        Double[] doubles = { 3.14, 2.71, 1.41 };
        System.out.println("Max Double: " + findMaximum(doubles));  // 3.14

        Long[] longs = { 100L, 50L, 200L };
        System.out.println("Max Long: " + findMaximum(longs));  // 200
    }
}

File: exam_prep/IST603/assignment/part4/FindMaximum.py
"""
Generic method find_maximum() with bounded type parameters.
Accepts a sequence of numbers (int, float, etc.) and returns the maximum value.
Uses typing.TypeVar(bound=...) to restrict to numeric types.
"""

from numbers import Real
from typing import TypeVar

T = TypeVar("T", bound=Real)


def find_maximum(array: list[T]) -> T:
    """
    Bounded type parameter T: must be a subtype of Real (int, float, etc.),
    so only numeric types are allowed. Type checkers will reject non-numeric types.
    """
    if not array:
        raise ValueError("Array must not be empty")
    maximum = array[0]
    for i in range(1, len(array)):
        if array[i] > maximum:
            maximum = array[i]
    return maximum


def main() -> None:
    ints = [3, 7, 2, 9, 1]
    print("Max int:", find_maximum(ints))  # 9

    floats = [3.14, 2.71, 1.41]
    print("Max float:", find_maximum(floats))  # 3.14

    mixed_numeric = [10, 5.5, 20]  # int and float both Real
    print("Max (mixed):", find_maximum(mixed_numeric))  # 20


if __name__ == "__main__":
    main()
